File: 2976-minimum-cost-to-convert-string-i/2976-minimum-cost-to-convert-string-i.go
func minimumCost(source string, target string, original []byte, changed []byte, cost []int) int64 {
    const INF = 1_000_000_007

	dist := [26][26]int64{}
	for i := 0; i < 26; i++ {
		for j := 0; j < 26; j++ {
			if i == j {
				dist[i][j] = 0
				continue
			}
			dist[i][j] = INF
		}
	}

	for i := 0; i < len(original); i++ {
		u, v, c := int(original[i]-'a'), int(changed[i]-'a'), int64(cost[i])
		dist[u][v] = min(dist[u][v], c)
	}

	// floyd-warshall
	for k := 0; k < 26; k++ {
		for i := 0; i < 26; i++ {
			for j := 0; j < 26; j++ {
				dist[i][j] = min(dist[i][j], dist[i][k]+dist[k][j])
			}
		}
	}

	ans := int64(0)
	for i := 0; i < len(source); i++ {
		s, t := int(source[i]-'a'), int(target[i]-'a')
		if dist[s][t] == INF {
			return -1
		}
		ans += dist[s][t]
	}

	return ans
}

// array, floyd-warshall, graph, string
// time: O(n)
// space: O(1)

File: 2976-minimum-cost-to-convert-string-i/2976-minimum-cost-to-convert-string-i.py
class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        dist = [[inf] * 26 for _ in range(26)]
        for i in range(len(dist)):
            dist[i][i] = 0
        
        for i in range(len(original)):
            s, t, c = ord(original[i]) - ord('a'), ord(changed[i]) - ord('a'), cost[i] # source, target, cost
            dist[s][t] = min(dist[s][t], c)
        
        for k in range(26):
            for i in range(26):
                for j in range(26):
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
        
        ans = 0
        for i in range(len(source)):
            if source[i] == target[i]:
                continue
            
            s, t = ord(source[i]) - ord('a'), ord(target[i]) - ord('a')
            if dist[s][t] == inf:
                return -1
            ans += dist[s][t]

        return ans

# array, floyd-warshall, graph, string
# time: O(n)
# space: O(1)